src/DynamoCore/Models/WorkspaceModel.cs

```diff
@@ -788,8 +796,87 @@ internal void IncrementPasteOffset()
+        /// <param name="description">a description of what the state does</param>
+        /// <param name="currentSelection">a set of NodeModels that are to be serialized in this state</param>
+        /// <param name="id">a GUID id for the state, if not supplied, a new GUID will be generated, cannot be a duplicate</param>
+        private void addState(string name, string description, IEnumerable<NodeModel> currentSelection, Guid id = new Guid())
+        {
+            if (currentSelection == null || currentSelection.Count() < 1)
+            {
+                throw new ArgumentException("currentSelection is empty or null");
+            }
+            var inputs = currentSelection;
+
+            if (Presets.Any(x => x.Guid == id))
+            {
+                throw new ArgumentException("duplicate id in collection");
+            }
+
+            var newstate = new PresetModel(name, description, inputs, id);
+            presets.Add(newstate);
+        }
+
+        public void RemoveState(PresetModel state)
+        {
+            if (Presets.Contains(state))
+            {
+                presets.Remove(state);
+            }
+        }
+
+        internal void ApplyPreset(PresetModel state)
+        {
+            if (state == null)
+            {
+                Log("Attempted to apply a PresetState that was null");
+                return;
+            }
+            //start an undoBeginGroup
+            using (var undoGroup = this.undoRecorder.BeginActionGroup())
+            {
+               //reload each node, and record each each modification in the undogroup
+                foreach (var node in state.Nodes)
+                {
+                    //check that node still exists in this workspace, 
+                    //otherwise bail on this node
+                    if (nodes.Contains(node))
+                    {
+                        var originalpos = node.Position;
+                        var serializedNode = state.SerializedNodes.ToList().Find(x => Guid.Parse(x.GetAttribute("guid")) == node.GUID);
+                        //overwrite the xy coords of the serialized node with the current position, so the node is not moved
+                        serializedNode.SetAttribute("x", originalpos.X.ToString());
+                        serializedNode.SetAttribute("y", originalpos.Y.ToString());
+
+                        this.undoRecorder.RecordModificationForUndo(node);
+                        this.ReloadModel(serializedNode);
+                    }
+                }
+                //select all the modified nodes in the UI
+                DynamoSelection.Instance.ClearSelection();
+                foreach(var node in state.Nodes)
+                {
+                    DynamoSelection.Instance.Selection.Add(node);
+                }
+            }
+        }
+        internal void CreatePresetStateFromSelection(string name, string description, List<Guid> IDSToSave)
```

Comment:
As this doesn't really use the selection, shouldn't this be called simply `AddPreset`?